scripts/cmake/vcpkg_configure_meson.cmake

```
@@ -54,12 +54,15 @@ function(vcpkg_internal_meson_generate_native_file _additional_binaries) #https:
     foreach(prog IN LISTS compiler)
         if(VCPKG_DETECTED_CMAKE_${prog}_COMPILER)
             string(REPLACE "CXX" "CPP" mesonprog "${prog}")
+            string(REPLACE "RC" "windres" mesonprog "${mesonprog}") # https://mesonbuild.com/Windows-module.html
             string(TOLOWER "${mesonprog}" proglower)
             string(APPEND NATIVE "${proglower} = '${VCPKG_DETECTED_CMAKE_${prog}_COMPILER}'\n")
         endif()
     endforeach()
-    if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS)
+    if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_DETECTED_CMAKE_C_COMPILER MATCHES "gcc|icc")
```

I don't understand this; could you explain? (preferably add a comment too?)

---

For gcc and ICC you can not explicitly set the linker

---

@autoantwort Can you add a comment in the code that explain this?

---

https://github.com/mesonbuild/meson/issues/8647#issuecomment-878673456  
I can do it next week

---

ok :)
Thank you!

---

It seems like the correct fix is to get the program name and pass that here rather than ignoring the setting completely?

---

When I understand it correctly you select the linker backend by adding a compiler flag and not by changing the linker executable. So the linker executeable is always the same and we can ignore the Program name. And the flag should be passed by VCPKG_LINKER_FLAGS or so

---

I guess, to put it another way, the right thing to do when we don't know how to attach a user setting isn't "just ignore the setting". Either we should make it work, or check that it's what we expect and fail if it isn't.

---

> check that it's what we expect and fail if it isn't.

Ok I will do that.

---

Hm, the current solution does not work. The linker is kind of baked in the compiler, but it is still a different executable. So the current check fails because of the following paths are different. 
```
  VCPKG_DETECTED_CMAKE_C_COMPILER=/usr/bin/x86_64-w64-mingw32-gcc
  VCPKG_DETECTED_CMAKE_LINKER=/usr/bin/x86_64-w64-mingw32-ld
```

---

Hm it seems that `-fuse-ld` is a linker flag

---

If I understand correctly you can not set the linker executable to something different for gcc. You have to use `-fuseld`. So we don't have to check if the linker executable was set to something different

---

The point is that if the user has attempted to change the linker in the toolchain or triplet and we hit a place that can't be applied, we should fail with a message saying that the customization was not applied rather than silently succeeding with something the customer did not expect.

---

But how you would set a different linker in the toolchain/triplet? Afaik you have to set the `-fuseld` flag and flags should then be passed to cmake.

---

> But how you would set a different linker in the toolchain/triplet? Afaik you have to set the `-fuseld` flag and flags should then be passed to meson.

To my understanding, they can change the linker by using `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` ( https://github.com/microsoft/vcpkg/blob/master/docs/users/triplets.md#vcpkg_chainload_toolchain_file ) and doing anything that would influence cmake to choose a different linker.

---

In the toolchain you would then do something like 
```cmake
set(CMAKE_SHARED_LINKER_FLAGS_INIT "-fuse-ld=lld")
set(CMAKE_STATIC_LINKER_FLAGS_INIT "-fuse-ld=lld")
```
to change the linker for gcc. The flags gets detected and written to `VCPKG_DETECTED_CMAKE_XXX_LINKER_FLAGS` and then passed to meson.